bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/geometry_collector.cpp

```diff
@@ -79,94 +81,203 @@ repo::core::model::RepoNodeSet repo::manipulator::modelconvertor::odaHelper::Geo
+	}
+
+	auto meshData = startOrContinueMeshByFormat(getMeshFormat(false, false, vertices.size()));
+
+	repo_face_t face;
+	for (auto i = 0; i < vertices.size(); ++i) 
+	{
+		auto& v = vertices[i];
+		int vertIdx = 0;
+
+		// match the vertex to generate an index for this face
+		if (meshData->vToVIndex.find(v) == meshData->vToVIndex.end())
+		{
+			vertIdx = meshData->rawVertices.size();
+			meshData->rawVertices.push_back(v);
+
+			meshData->vToVIndex.insert(
+				std::pair<repo::lib::RepoVector3D64,
+				std::pair<int, repo::lib::RepoVector3D64>>(
+					v,
+					std::pair<int,
+					repo::lib::RepoVector3D64>(
+						vertIdx,
+						repo::lib::RepoVector3D64())));
+
+
+			// if the vertex has not yet been seen, also update the bounding box
+
+			if (!meshData->boundingBox.size())
+			{
+				meshData->boundingBox.push_back({ (float)v.x, (float)v.y, (float)v.z });
+				meshData->boundingBox.push_back({ (float)v.x, (float)v.y, (float)v.z });
+			}
+			else
+			{
+				meshData->boundingBox[0][0] = meshData->boundingBox[0][0] > v.x ? (float)v.x : meshData->boundingBox[0][0];
+				meshData->boundingBox[0][1] = meshData->boundingBox[0][1] > v.y ? (float)v.y : meshData->boundingBox[0][1];
+				meshData->boundingBox[0][2] = meshData->boundingBox[0][2] > v.z ? (float)v.z : meshData->boundingBox[0][2];
+
+				meshData->boundingBox[1][0] = meshData->boundingBox[1][0] < v.x ? (float)v.x : meshData->boundingBox[1][0];
+				meshData->boundingBox[1][1] = meshData->boundingBox[1][1] < v.y ? (float)v.y : meshData->boundingBox[1][1];
+				meshData->boundingBox[1][2] = meshData->boundingBox[1][2] < v.z ? (float)v.z : meshData->boundingBox[1][2];
+			}
+
+			if (minMeshBox.size()) {
+				minMeshBox[0] = v.x < minMeshBox[0] ? v.x : minMeshBox[0];
+				minMeshBox[1] = v.y < minMeshBox[1] ? v.y : minMeshBox[1];
+				minMeshBox[2] = v.z < minMeshBox[2] ? v.z : minMeshBox[2];
+			}
+			else {
+				minMeshBox = { v.x, v.y, v.z };
+			}
+		}
+		{
+			auto iterators = meshData->vToVIndex.equal_range(v);
+			vertIdx = iterators.first->second.first;
+		}
+
+		face.push_back(vertIdx);
+	}
+
+	meshData->faces.push_back(face);
 }
 
 void GeometryCollector::addFace(
 	const std::vector<repo::lib::RepoVector3D64> &vertices,
 	const repo::lib::RepoVector3D64& normal,
 	const std::vector<repo::lib::RepoVector2D>& uvCoords)
 {
-	if (!meshData.size()) startMeshEntry();
-
-	bool currentEntryHasUV = currentEntry->uvCoords.size();
-	bool faceHasUV = uvCoords.size();
-	if (currentEntry->rawVertices.size() && currentEntryHasUV != faceHasUV) {
-		//Start a new mesh if the current mesh has UVs but this face doesn't or vice versa.
-		auto lastMeshName = nextMeshName;
-		stopMeshEntry();
-		nextMeshName = lastMeshName + "_";
-		repoDebug << "Staring a new mesh due to UV difference";
-		startMeshEntry();
+	if (!vertices.size())
+	{
+		repoError << "Vertices size[" << vertices.size() << "] is unsupported. A face must have more than 0 vertices";
+		exit(-1);
```

Comment:
ideally if we can do some error handling that gets propagate all the way back up that'll be great.

Reason being the bouncer_worker takes the error code from the program and it's base on https://github.com/3drepo/3drepobouncer/blob/master/bouncer/src/repo/error_codes.h

Using those will give us a better idea of why something failed (which ultimately propagates all the way back up to the web UI to show the user)

Reply:
@carmenfan OK, so shall I update this exit call to return an appropriate error code, or throw an exception that gets caught further up (or similar)? 
(I was wondering about this as ODA (I am fairly sure it comes from ODA) throws a "exited with fatal code" exception during teardown when the debugger is attached, even though the exit code is 0!)

Reply:
Whatever works and is easiest is fine!

Reply:
Seems like no matter what we do ODA intercepts our exit code or exception, and re-issues it with its own that obscures ours. This might be because the DataProcessor is a "device" to ODA, so ODA is designed not to allow it to throw or otherwise try and take down the process.
In any case, I've added an error code member to the geometry collector that is checked by OdaModelImport.